File: 155. Min Stack.cpp
#include <iostream>
#include <math.h>
#include <algorithm>
#include <string>
#include <list>
#include <vector>
#include <set>
#include <unordered_set>
#include <map>
#include <unordered_map>
#include <queue>
#include <stack>
#include <limits>
#include <thread>
#include <mutex>
#include <fstream>
#include <memory>
using namespace std;
class MinStack {
private:
	std::stack<int> data;
	std::stack<int> minData;
	
public:
	/* initialize your data structure here. */
	MinStack() {
		data = stack<int>();
		minData = stack<int>();
	}
	
	void push(int x) {
		if (minData.empty() || x <= minData.top()) {
			minData.push(x);
		}
		
		data.push(x);
	}
	
	void pop() {
		if (data.top() == minData.top()) {
			data.pop();
			minData.pop();
		}
		else {
			data.pop();
		}
	}
	
	int top() {
		return data.top();
	}
	
	int getMin() {
		return minData.top();
	}
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main(int argc, char *argv[]) {
	
}

File: 155. Min Stack.java
import java.util.*;

class MinStack {

	public Stack<Integer> minStk;
	public Stack<Integer> stk;
	
	/** initialize your data structure here. */
	public MinStack() {
		minStk = new Stack<Integer>();
		stk = new Stack<Integer>();
	}
	
	public void push(int x) {
		stk.push(x);
		if (minStk.empty() || minStk.peek() >= x) {
			minStk.push(x);
		}
	}
	
	public void pop() {
		if (minStk.peek().equals(stk.peek())){
			minStk.pop();
		}
		
		stk.pop();
	}
	
	public int top() {
		return stk.peek();
	}
	
	public int getMin() {
		return minStk.peek();
	}
}

class a {
	public static void main(String[] args) {
		MinStack obj = new MinStack();
		obj.push(512);
		obj.push(-1024);
		obj.push(-1024);
		obj.push(512);
		
		System.out.println(Arrays.toString(obj.minStk.toArray()));
		System.out.println(Arrays.toString(obj.stk.toArray()));
		
		obj.pop();
		System.out.println(Arrays.toString(obj.minStk.toArray()));
		System.out.println(Arrays.toString(obj.stk.toArray()));
		System.out.println( obj.getMin() );
		
		obj.pop();
		System.out.println(Arrays.toString(obj.minStk.toArray()));
		System.out.println(Arrays.toString(obj.stk.toArray()));
		System.out.println( obj.getMin() );
		
		
		obj.pop();
		System.out.println(Arrays.toString(obj.minStk.toArray()));
		System.out.println(Arrays.toString(obj.stk.toArray()));
		System.out.println( obj.getMin() );
	}
}